### PR TITLE
Dialog Editor: Updated Styles

### PR DESCRIPTION
**Pull request by dtaylor113**
Updated the Dialog Editor:

- Updated layout:  
    - Dialog Name and Description as horizontal form 
    - Toolbox on left, Tabs/Sections/Fields on right. 
image
- Added grab handles to toolbox items, boxes/sections, and fields/elements (appear on hover):
Toolbox Item:
image
Section:
image
Field:
image
- Added edit/pencil and remove/trashcan icons to Tabs, Sections, and Fields (appear on hover):
image
image
image
- Details Modal:
    - Added component specific modal Titles Ex: "Edit Textarea Field" and "Edit Tab Details"
    - Removed Delete button
- Replaced "box" terminology with "section"

Pivotal: https://www.pivotaltracker.com/n/projects/1613907/stories/150461077
@romanblanco 
@serenamarie125

### CODE REVIEW

**Comment by martinpovolny** [comment]
@serenamarie125 : is this ok?

**Comment by martinpovolny** [comment]
@romanblanco : any comments?

File: demo/controllers/dialogEditorController.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ export default class DialogEditorController {
           'label': 'New tab',
```
Comment (serenamarie125) [inline demo/controllers/dialogEditorController.ts]: @dtaylor113 can you change this to "New Tab" ( capital T ) ?

**Comment by martinpovolny** [comment]
> @dtaylor113 Demo page works, but the editor in Automate -> Customization seems to be missing some CSS styles:

@romanblanco, @dtaylor113 : can you, please, figure out what style is missing so that we can add it to the OPS UI if needed?

**Comment by dtaylor113** [comment]
Hi, working with @romanblanco, we identified the file which needs to be updated in ui-classic: ```app/views/miq_ae_customization/editor.html.haml```.  This file needs to match ```demo/views/dialog/editor.html``` in this PR.   Since I do not have manageiq nor ui-classic setup, and am not familiar with .haml files;  @romanblanco has graciously agreed to update the template in ui-classic  -thanks @romanblanco :-)

File: demo/controllers/dialogEditorController.ts (hunk 1 above)
Comment (dtaylor113) [reply demo/controllers/dialogEditorController.ts]: Ok, fixed

**Comment by romanblanco** [comment]
@dtaylor113 @skateman The PR in `manageiq-ui-classic` is ready, but I've probably found one more issue with toolbar icons. Can you look at it?

File: src/styles/dialog-editor-boxes.scss [hunk 2]
```diff
@@ -0,0 +1,151 @@
+.selected {
+  border: 3px solid $color-blue-300;
+}
+
+&.disabled {
+  color: $color-gray;
+  font-size: 18px;
```
Comment (skateman) [inline src/styles/dialog-editor-boxes.scss]: @dtaylor113 this is the conflicting line

Comment (dtaylor113) [reply src/styles/dialog-editor-boxes.scss]: Hi,
I removed the conflicting line, please retest.
Thanks,
Dave

File: src/styles/dialog-editor-boxes.scss [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  img {
+    height: 100%;
+    margin-right: 8px;
+    object-fit: contain;
+  }
+}
+
+.form-group {
```
Comment (skateman) [inline src/styles/dialog-editor-boxes.scss]: Don't we want to namespace everything in the CSS, I'm pretty sure that the `.form-group` can be also conflicting. Maybe not now, but what if someone just adds something here without thinking. cc @karelhala @himdel

Comment (skateman) [reply src/styles/dialog-editor-boxes.scss]: We're in a component (custom element) that can be used as a namespace.

File: src/dialog-editor/services/modal/modalService.ts [hunk 4]
```diff
@@ -81,6 +85,40 @@ class ModalController {
...
+              titleLabel = "Datepicker";
+              break;
+            case 'DialogFieldDateTimeControl':
+              titleLabel = "Timepicker";
+              break;
+            case 'DialogFieldTagControl':
+              titleLabel = "Tag Control";
+              break;
+          }
+          this.modalTitle =  __("Edit " + titleLabel + " Field");
```
Comment (himdel) [inline src/dialog-editor/services/modal/modalService.ts]: broken i18n .. you can *never* use + inside of `__`

File: src/styles/dialog-editor-boxes.scss (hunk 3 above)
Comment (himdel) [reply src/styles/dialog-editor-boxes.scss]: Definitely need to namespace everything. No way can we redefine `form-group` or similar without breaking everything in manageiq :) 

I think in the long run, we should remove everything from `src/styles/` in favor of per-component styles - which webpack would namespace by component automagically. But for now, something like `.dialog-editor .form-group` *might* suffice.

Comment (dtaylor113) [reply src/styles/dialog-editor-boxes.scss]: Ok, removed ```.form-group```.  It wasn't being used and was an old artifact.  -thanks

File: src/dialog-editor/services/modal/modalService.ts (hunk 4 above)
Comment (dtaylor113) [reply src/dialog-editor/services/modal/modalService.ts]: Ok, fixed the i18n issues.  -thanks

File: src/styles/dialog-editor-boxes.scss (hunk 3 above)
Comment (dtaylor113) [reply src/styles/dialog-editor-boxes.scss]: Reviewed the new .scss files added in this PR; everything appears to be namespaced or at least have unique parent class names.

File: src/dialog-editor/services/modal/modalService.ts [hunk 5]
```diff
@@ -81,6 +85,40 @@ class ModalController {
...
+              titleLabel = __("Datepicker");
+              break;
+            case 'DialogFieldDateTimeControl':
+              titleLabel = __("Timepicker");
+              break;
+            case 'DialogFieldTagControl':
+              titleLabel = __("Tag Control");
+              break;
+          }
+          this.modalTitle =  __("Edit ") + titleLabel +  __(" Field");
```
Comment (karelhala) [inline src/dialog-editor/services/modal/modalService.ts]: This is one way of creating strings in JS, however consider using [Template_literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals).
```Typescript
__(`Edit ${titleLabel} Field`);
```

File: src/dialog-editor/services/modal/modalService.ts [hunk 6]
```diff
@@ -81,6 +85,40 @@ class ModalController {
             }
           );
         }
+        // set modal title
+        if (!this.modalData.dynamic) {
+          var titleLabel;
+
+          switch (this.modalData.type) {
```
Comment (karelhala) [inline src/dialog-editor/services/modal/modalService.ts]: Personally I don't like switches. They are too big, hard to read and hard to interpret. I prefer something which is (at least for me) better to read.
```Typescript
const titles = {
  DialogFieldTextBox:         __('Text Box'),
  DialogFieldTextAreaBox:     __('Text Area'),
  DialogFieldCheckBox:        __('Check Box'),
  DialogFieldDropDownList:    __('Dropdown'),
  DialogFieldRadioButton:     __('Radio Button'),
  DialogFieldDateControl:     __('Datepicker'),
  DialogFieldDateTimeControl: __('Timepicker'),
  DialogFieldTagControl:      __('Tag Control')
}
const titleLabel = this.modalData.type in titles && titles[this.modalData.type];
this.modalTitle =  __(`Edit ${titleLabel} Field`);
```

File: src/dialog-editor/components/field/fieldComponent.ts [hunk 7]
```diff
@@ -34,6 +36,25 @@ class FieldController {
...
+  /**
+   * Remove Field
+   * @memberof FieldController
+   * @function remmoveField
+   * @param {number} tabId is an index of tab, where the box is placed
+   * @param {number} boxId is an index of box, where the field is placed
+   * @param {number} fieldId is an index of field
+   */
+  public removeField(tabId: number, boxId: number, fieldId: number) {
+    _.remove(
```
Comment (karelhala) [inline src/dialog-editor/components/field/fieldComponent.ts]: This construction begs for undefined. Can you rewrite it perhaps to this?
```Typescript
class getGroupByBoxId {
  //..
  private static getTabById(tabs, tabId) {
    return tabs[tabId];
  }
  private static getGroupByBoxId(groups, boxId) {
    return groups[boxId];
  }

  private static getDialogFields(tabs, tabId, boxId) {
    const tab = FieldController.getTabById(tabs, tabId);
    const group =  tab && FieldController.getGroupByBoxId(tab.dialog_groups, boxId);
    return group && group.dialog_fields;
  }

  public removeField(tabId: number, boxId: number, fieldId: number) {
    const fields = FieldController.getDialogFields(this.DialogEditor.getDialogTabs(), tabId, boxId);
    fields && _.remove(fields, (field: any) => field.position === fieldId);
  }
}
```

File: src/dialog-editor/components/box/box.html [hunk 8]
```diff
@@ -44,15 +45,16 @@
...
         <i class="fa fa-plus-circle"></i>
       </div>
-      <h1 translate>Start with adding a box</h1>
+      <h1 translate style="cursor: pointer;">Start with adding a section</h1>
     </div>
-    <div class="panel panel-default nosort">
+    <div class="add-section-box nosort">
       <a ng-click='vm.addBox()' translate>
-        <i class="pficon-add-circle-o"></i>&nbsp;Create Box
+        <i class="pficon-add-circle-o"></i>&nbsp;&nbsp;Add section
```
Comment (karelhala) [inline src/dialog-editor/components/box/box.html]: Carefull not translated.
```html
<i class="pficon-add-circle-o"></i>&nbsp;&nbsp;{{__('Add section')}}
```

File: src/dialog-editor/services/modal/modalService.ts (hunk 6 above)
Comment (dtaylor113) [reply src/dialog-editor/services/modal/modalService.ts]: Hi @karelhala, there are 3 other ```switch``` statements in this file, so I think in this case it's more consistent to use a ```switch```

File: src/dialog-editor/services/modal/modalService.ts (hunk 5 above)
Comment (dtaylor113) [reply src/dialog-editor/services/modal/modalService.ts]: Hi, I'm not familiar enough with your translation tools.  I wasn't allowed to use a '+' in there, so I'm hesitant to use a template literal.

File: src/dialog-editor/components/box/box.html (hunk 8 above)
Comment (dtaylor113) [reply src/dialog-editor/components/box/box.html]: Hi @karelhala, fixed this last translation issue.  Thanks for your review comments :-)

File: src/dialog-editor/components/field/fieldComponent.ts (hunk 7 above)
Comment (dtaylor113) [reply src/dialog-editor/components/field/fieldComponent.ts]: This class and methods are long standing, I do not want to start making changes in there, especially if unrelated to this particular PR.

**Comment by dtaylor113** [comment]
> please change all double quotes to single quotes.

Done :-)

File: src/dialog-editor/services/modal/modalService.ts (hunk 6 above)
Comment (romanblanco) [reply src/dialog-editor/services/modal/modalService.ts]: @karelhala i like the changes you've suggested, but it's out of scope for this PR, I'll create a new PR with changes you've suggested.

**Comment by martinpovolny** [comment]
@skateman : please, confirm that all the styling issues have been fixed and we can merge this one.

File: src/dialog-editor/services/modal/modalService.ts [hunk 9]
```diff
@@ -81,6 +85,40 @@ class ModalController {
             }
           );
         }
+        // set modal title
+        if (!this.modalData.dynamic) {
+          var titleLabel;
```
Comment (skateman) [inline src/dialog-editor/services/modal/modalService.ts]: use `let` instead of `var`

**Comment by martinpovolny** [comment]
Ok, if that's the last bit then let's do it in a separate PR. Thx!

File: src/dialog-editor/services/modal/modalService.ts (hunk 9 above)
Comment (romanblanco) [reply src/dialog-editor/services/modal/modalService.ts]: I'll fix this one in next PR together with [other changes](https://github.com/ManageIQ/ui-components/pull/139#discussion_r137289809)

**Pull request merged**